### PR TITLE
SslHandler.wrap(...) must ensure it not loose the original exception …

### LINKED ISSUE #9971: SslHandler.flush throws IllegalArgumentException on upstream write exception
### Expected behavior

Did not expect flush to throw IllegalArgumentException

### Actual behavior
```
promise already done: DefaultChannelPromise@3d07c0a4(
io.netty.channel.AbstractChannelHandlerContext.isNotValidPromise - Line 910 (AbstractChannelHandlerContext.java)
io.netty.channel.AbstractChannelHandlerContext.write - Line 714 (AbstractChannelHandlerContext.java)
io.netty.handler.ssl.SslHandler.finishWrap - Line 850 (SslHandler.java)
io.netty.handler.ssl.SslHandler.wrap - Line 836 (SslHandler.java)
io.netty.handler.ssl.SslHandler.wrapAndFlush - Line 753 (SslHandler.java)
io.netty.handler.ssl.SslHandler.flush - Line 734 (SslHandler.java)
io.netty.channel.AbstractChannelHandlerContext.invokeFlush0 - Line 776 (AbstractChannelHandlerContext.java)
io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush - Line 802 (AbstractChannelHandlerContext.java)
io.netty.channel.AbstractChannelHandlerContext.write - Line 814 (AbstractChannelHandlerContext.java)
io.netty.channel.AbstractChannelHandlerContext.writeAndFlush - Line 794 (AbstractChannelHandlerContext.java)
```

### Steps to reproduce

Occurs when an upstream handler throws an exception in write()

### Netty version

4.1.22-current

### JVM version (e.g. `java -version`)
1.8
### OS version (e.g. `uname -a`)
linux

### PR DESCRIPTION
**Pull request by normanmaurer**
…when finishWrap(...) fails

Motivation:

When SslHandler.finishWrap throws an exception, ensure that the promise and buf is not reused to avoid throwing IllegalArgumentException or IllegalReferenceCountException which causes the original exception to be lost.

Modification:

The change ensures that the values for the promise and bytebuf are nulled before calling finishWrap so that it will not be called again with the same arguments.

Result:

Fixes #9971 .

Co-authored-by: Norman Maurer <norman_maurer@apple.com>

### CODE REVIEW

**Comment by normanmaurer** [comment]
@atcurtis PTAL

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 1]
```diff
@@ -858,11 +858,25 @@ private void wrap(ChannelHandlerContext ctx, boolean inUnwrap) throws SSLExcepti
...
-                            out = null;
+                            final ByteBuf b;
+
+                            if (out.isReadable()) {
+                                // There is something in the out buffer. Ensure we null it out so it is not re-used.
+                                b = out;
+                                out = null;
+                            } else {
+                                // If out is not readable we can re-use it and so save an extra allocation
+                                b = null;
```
Comment (njhill) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: How about `out.clear()` here too?

Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: Good idea. Let me do this.

Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: this should not be needed.

**Pull request merged**